Move in place three sections to docs "How-to" (in Managing your installation)

giovannipizzi opened the pull request:
In particular, moving:

how-to:tune performance (fixes #4002)
how-to:update installation (fixes #4003)
how-to:backup (fixes #4004)

I'm doing these in a single commit since they are closely related (especially for what concerns links)

giovannipizzi committed:
Moving in the new documentation scaffolding structure the following sections:

how-to:tune performance (fixes #4002)
how-to:update installation (fixes #4003)
how-to:backup (fixes #4004)

I'm doing these in a single commit since they are closely related (especially for what concerns links)

giovannipizzi commented:
The only question I have is the folder `installation_more` where I put additional files that I think should be in a different page, otherwise the page becomes too long with not-so-critical details.

Suggestions for a different name (or different location) are welcome.

csadorf commented on docs/source/howto/installation_more/index.rst:
```diff
@@ -62,3 +61,14 @@ Finally, check that the data directory has indeed changed::
 
 If everything went fine, you can delete the old database location.
+
+.. _how-to:installation:more:disable_repo_indexing:
+
+#############################################
+Exclude file repository from ``locate`` index
+#############################################
+
+Many Linux distributions include the ``locate`` command to quickly find files and folders, and run a daily cron job ``updatedb.mlocate`` to create the corresponding index.
+A large file repository can take a long time to index, up to the point where the hard drive is constantly indexing.
+
+In order to exclude the repository folder from indexing, add its path to the ``PRUNEPATH`` variable in the ``/etc/updatedb.conf`` configuration file (use ``sudo``).
```
We can place this content within a dropdown close to the related note.

giovannipizzi replied:
Ok

csadorf reviewed (changes requested):
I made some general suggestions on how to improve the style and maybe make the content a bit more concise.

For some of the content, especially the backup section with O/S specific instructions, I make suggestions on how to either move or place it such that it does not impede the general flow and is better separated from content that is more generally applicable. I think we should aim at streamlining these sections such that they remain clearly problem-oriented. See my specific comments for details.

I also suggest to introduce a new top-level how-to section on "how to run on clusters". While very relevant for users that do run on clusters, I think it is obstructive for those that don't, so a better separation is probably desirable. It will also help cluster users find the related content more easily.

chrisjsewell commented on docs/source/install/installation.rst:
```diff
@@ -170,11 +170,11 @@ If you use the same names as in the example commands above, then during the ``ve
 
-.. note:: Do not forget to backup your database (instructions :ref:`here<backup_postgresql>`).
+.. note:: Do not forget to backup your database (instructions :ref:`here<how-to:installation:backup:postgresql>`).
 
 .. note:: If you want to move the physical location of the data files
   on your hard drive AFTER it has been created and filled, look at the
-  instructions :ref:`here<move_postgresql>`.
+  instructions :ref:`here<how-to:installation:more:move_postgresql>`.
```
This file is removed in #4060, so probably best to add these links in the new sections there

giovannipizzi replied:
Thanks @chrisjsewell - I guess you are working on #4060? Could you fix the links there already? Thanks!

chrisjsewell replied:
yep I am, and will do

giovannipizzi commented on docs/source/howto/data.rst:
```diff
@@ -244,7 +244,7 @@ However, storing large amounts of data within the database comes at the cost of
 A data type may safely use both the database and file repository in parallel for individual properties.
 Properties stored in the database are stored as *attributes* of the node.
-The node class has various methods to set these attributes, such as :py:`~aiida.orm.node.Node.set_attribute` and :py:`~aiida.orm.node.Node.set_attribute_many`.
+The node class has various methods to set these attributes, such as :py:meth:`~aiida.orm.node.Node.set_attribute` and :py:meth:`~aiida.orm.node.Node.set_attribute_many`.
```
Yes, but I spotted that it was rendering incorrectly and I fixed it. BTW, the code path is also wrong - if OK to fix in this commit, I would fix that as well.

csadorf replied:
I'd prefer to keep unrelated changes in a separate change set.

giovannipizzi replied:
Ok, I'll revert. For your ref, here is the correct string to use:
```rst
:py:meth:`~aiida.orm.nodes.node.Node.set_attribute` and :py:meth:`~aiida.orm.nodes.node.Node.set_attribute_many`
```

giovannipizzi committed:
Addressing review comments

Some reorganisation of content and some rewording

Co-authored-by: Carl Simon Adorf <carl.simon.adorf@gmail.com>

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+     :ref:`Disable this<how-to:installation:more:disable_repo_indexing>`.
+
+  4. The verdi deamon can manage an arbitrary number of parallel workers; by default only one is activated. If ``verdi daemon status`` shows the daemon worker(s) constantly at high CPU usage, use ``verdi daemon incr X`` to add ``X`` daemon workers.
+     However, don't use many more workers than CPU cores on your machine (or ideally, if you have many cores, leave one or two available for the database).
```
I adapted your suggestion to still mention to use 1/2 cores less (in @sphuber results, this was important)

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+
+5. Update your ``aiida-core`` installation.
+
+    * If you have installed AiiDA through ``pip`` simply run: ``pip install --upgrade aiida-core``.
```
Ok. Do you know what is the correct conda command to use?

giovannipizzi replied:
Ok, I'll take care of this.

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+5. Update your ``aiida-core`` installation.
+
+    * If you have installed AiiDA through ``pip`` simply run: ``pip install --upgrade aiida-core``.
+    * If you have installed from the git repository using ``pip install -e .``, first delete all the ``.pyc`` files (``find . -name "*.pyc" -delete``) before updating your branch.
```
This has caused a lot of pain in the past (@sphuber can confirm), so I would vote for keeping this (unless we are sure that in python 3 the problem is solved, but I'm not sure it is). If you feel we should mention `__pycache__`, it's ok with me (should we just delete those as well?). But I think that if we delete the *.pyc, in py3 these are inside the `__pycache__` folder and the suggestion would still work.

In detail, the reason for the problems that you get if you don't delete them is when you delete or rename a file. In particular, this happened a lot in the 0.x->1.x transition (might be less relevant in the 1.x versions, but we might still end up doing some refactoring of internal python files). However, since *.pyc are not tracked by git, git would leave them at their old location together with their folders, and python would happily continue to load them.

giovannipizzi commented:
Thanks @csadorf, I'll apply the rest of the comments (mostly involving moving stuff around) tomorrow. Sorry also for the many commits, I didn't get at the beginning how to batch suggestions together (was working from the discussion pane and not from the files pane).
Anyway this will be squash-merged, or maybe when working on the rest I'll already rebase.

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -223,23 +223,269 @@ the configuration directory ``~/project_a/.aiida`` will be used.
-`#4004`_
+A full backup of an AiiDA instance and AiiDA managed data requires a backup of:
+
+ * the profile configuration in the ``config.json`` file located in the ``.aiida`` folder.
```
Check whether the bullet points are now showing (before they were somehow missing)

giovannipizzi replied:
It's ok now

ltalirz commented on docs/source/howto/installation.rst:
```diff
@@ -223,23 +223,269 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+Repository backup (``.aiida`` folder)
+-------------------------------------
+
+For **small repositories** back up the ``.aiida`` folder either by making a full copy or using tools for incremental backups like ``rsync``.
```
```suggestion
For **small repositories** with less than 100k nodes, simply back up the ``.aiida`` folder using standard backup software.
For example, the ``rsync`` utility supports incremental backups, and a backup command might look like ``rsync ...``
```

ltalirz replied:
It would be quite useful for people to already have the sensible flags here.

ltalirz reviewed (commented):
thanks a lot @giovannipizzi 

I added two comments I had written down for myself to fix at some point.
They are easy to address.

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+  4. The verdi deamon can manage an arbitrary number of parallel workers; by default only one is activated. If ``verdi daemon status`` shows the daemon worker(s) constantly at high CPU usage, use ``verdi daemon incr X`` to add ``X`` daemon workers.
+     However, don't use many more workers than CPU cores on your machine (or ideally, if you have many cores, leave one or two available for the database).
+
+  5. If you submit to a supercomputer shared by many users (e.g., in a supercomputer center), be careful not to overload the supercomputer with too many jobs:
```
Ok, I'll create a section "Running on supercomputers"

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+
+    The configuration step creates two files in the "backup folder": a ``backup_info.json`` configuration file (can also be edited manually) and a ``start_backup.py`` script.
+
+    .. note::
+
+        Using the same disk for your backup forgoes protection against the most common cause of data loss: disk failure.
+        One simple option is to use a destination folder mounted over ssh.
+
+        On Ubuntu, install ``sshfs`` using ``sudo apt-get install sshfs``.
+        Imagine you run your calculations on `server_1` and would like to back up regularly to `server_2`.
+        Mount a `server_2` directory on `server_1` using the following command on `server_1`:
+
+        .. code-block:: shell
+
+            sshfs -o idmap=user -o rw backup_user@server_2:/home/backup_user/backup_destination_dir/ /home/aiida_user/remote_backup_dir/
+
+        Use ``gnome-session-properties`` in the terminal to add this line to the actions performed at start-up.
+        Do **not** add it to your shell's startup file (e.g. ``.bashrc``) or your computer will complain that the mount point is not empty whenever you open a new terminal.
```
Ok, I'll use the panels (they will appear once #4107 is merged as well, the only needed change from there is to pip-install version 0.4.0 of sphinx_panels instead of 0.1.0)

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+
+Finally, if you have a separate automatic backup of your home directory set up, do not forget to exclude the repository folder.
+
+.. _how-to:installation:backup:repository:configuration_options:
+
+Configuration options
+^^^^^^^^^^^^^^^^^^^^^
```
Ok, will do.

giovannipizzi commented on docs/source/howto/installation_more/for_cluster_admins.rst:
```diff
@@ -0,0 +1,35 @@
+.. _how-to:installation:performance:for_cluster_admins:
```
Ok, will do (and keep it in a dropdown as it's only needed by very few people).

csadorf commented:
@giovannipizzi Is this ready for another round of review?

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+
+            00 03 * * * verdi -p PROFILENAME run /home/USERNAME/.aiida/backup/start_backup.py 2>&1 | mail -s "Incremental backup of the repository" USER_EMAIL@domain.net
+
+        This will launch the backup of the database every day at 3 AM (03:00), and send the output (or any error message) to the email address specified at the end (provided the ``mail`` command -- from ``mailutils`` -- is configured appropriately).
```
Ok

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -26,23 +26,271 @@ Configuring your installation
+.. _how-to:installation:backup:postgresql:
+
+Database backup
+---------------
```
I simplified the section and linked to PSQL docs

giovannipizzi commented:
Hi! 
I was just going to ping you, I just needed to do a force push because I noticed a small problem.
I also reorganised the commits, there are only two now.
I now addressed all comments (I think) and it's ready to be reviewed again! Thanks

giovannipizzi commented:
@csadorf FYI I'm going to add a new short section that was just contributed by Nicola Spallanzani, as it fits in here; I'm also going to rebase so you can check on CircleCI

giovannipizzi committed:
Adding section on how to use passphrase-protected SSH keys

Co-authored-by: Nicola Spallanzani <nicola.spallanzani@max-centre.eu>

giovannipizzi commented:
@csadorf OK, done!

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+    For a given computer, you can increase how many seconds must pass between requests.
+    First load the computer in a shell with ``computer = load_computer(<COMPUTER_NAME>)``.
+    You can check the current value in seconds (by default, 10) with ``computer.get_minimum_job_poll_interval()``.
+    You can then set it to a higher value using::
+
+      computer.set_minimum_job_poll_interval(<NEW_VALUE_SECONDS>)
```
We have decided in a separate discussion to always use explicit mark-up for code segments, see [here](https://github.com/aiidateam/aiida-core/wiki/Writing-documentation#code-segments).

giovannipizzi replied:
Ok, sorry, I had missed that - I'll fix this

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+.. _how-to:installation:running-on-supercomputers:
+
+Running on supercomputers
```
I'd recommend that we move this section a bit further down to after "Backing up your installation" given that this section is probably slightly less relevant for the average user in comparison to the other sections.

giovannipizzi replied:
Ok, will move

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+.. note::
+
+   The passphase is stored in the agent only until the next reboot. If you shut down or restart the AiiDA machine, before starting the AiiDA deamon remember to run the ``ssh-add`` command again.
```
```suggestion
   The passphase is stored in the agent only until the next reboot.
   If you shut down or restart the AiiDA machine, before starting the AiiDA deamon remember to run the ``ssh-add`` command again.
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+Make sure that this is the case also in newly opened bash shells.
+
+Then, when configuring the corresponding AiiDA computer (via ``verdi computer configure``), make sure to specify ``true`` to the question ``Allow ssh agent``.
+If you already configured the computer and just want to adapt the computer configuration, you can run again
```
```suggestion
If you already configured the computer and just want to adapt the computer configuration, just rerun
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+    verdi computer configure ssh COMPUTERNAME
+
+After the configuration, you should verify that AiiDA can connect to the computer successfully using:
```
```suggestion
After the configuration, you should verify that AiiDA can connect to the computer with:
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+If you submit to a supercomputer shared by many users (e.g., in a supercomputer center), be careful not to overload the supercomputer with too many jobs:
+
+  * keep the number of jobs in the queue under control (the exact number depends on the supercomputer: discuss this with your supercomputer administrators, and you can redirect them to :ref:`this page<how-to:installation:running-on-supercomputers:for_cluster_admins>` that may contain useful information for them).
```
```suggestion
  * limit the number of jobs in the queue (the exact number depends on the supercomputer: discuss this with your supercomputer administrators, and you can redirect them to :ref:`this page<how-to:installation:running-on-supercomputers:for_cluster_admins>` that may contain useful information for them).
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+If you submit to a supercomputer shared by many users (e.g., in a supercomputer center), be careful not to overload the supercomputer with too many jobs:
+
+  * keep the number of jobs in the queue under control (the exact number depends on the supercomputer: discuss this with your supercomputer administrators, and you can redirect them to :ref:`this page<how-to:installation:running-on-supercomputers:for_cluster_admins>` that may contain useful information for them).
+    While in the future `this might be dealt by AiiDA automatically <https://github.com/aiidateam/aiida-core/issues/88>`_,
```
```suggestion
    While in the future `this might be dealt with by AiiDA automatically <https://github.com/aiidateam/aiida-core/issues/88>`_,
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+Optimising the SLURM scheduler configuration
+--------------------------------------------
+
+If too many jobs are submitted at the same time to the queue, SLURM might have troubles in dealing with the new submissions.
```
```suggestion
If too many jobs are submitted at the same time to the queue, SLURM might have troubles in dealing with new submissions.
```

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+.. _how-to:installation:running-on-supercomputers:for_cluster_admins:
+
+Optimising the SLURM scheduler configuration
+--------------------------------------------
```
I'm wondering if we should move this content to the wiki instead and link there. It's informative in this context, but it does not really document the "AiiDA" software itself.

I am also a bit worried that presenting this information prominently in the documentation might cause an issue, because it is very assuming. I am thinking about admins of large supercomputer centers that have spent thousands of hours fine-tuning their scheduler systems, who might get a little weirded out (or worse: offended) by getting general recommendations on how to set up their system from AiiDA users. Our close relationship with CSCS is certainly not the norm for the average user.

giovannipizzi replied:
Ok, I see your point, so I moved to the wiki (but I would still link from here, as otherwise the information will be very hard to find).

Anyway, I'm not particularly worried of people getting offended in this case - we are not giving them recommendations, but just making public some hints provided to us by CSCS, that is one of the biggest HPC centres worldwide. Actually, if other centres have more suggestions, I'd be happy to extend that page.

csadorf replied:
Maybe "offended" wasn't the right word, but either way, the wiki is probably a better place for this.

csadorf commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,344 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+    The configuration step creates two files in the "backup folder": a ``backup_info.json`` configuration file (can also be edited manually) and a ``start_backup.py`` script.
+
+    .. dropdown:: Notes on using a SSH mount for the backups (on Linux)
```
Again, at least on my local computer, this doesn't work.

giovannipizzi replied:
See above

csadorf commented on docs/source/howto/load-singlesshagent.sh:
```diff
@@ -0,0 +1,68 @@
+# Make sure you source this script rather than executing it
```
Is this a new feature? Then we should probably not introduce it as part of the docs-revamp.

giovannipizzi replied:
It is not a new feature, but it is definitely new documentation. But anyway we are creating new docs for other how-to pages, and this is very timely considering the recent cyberattack (actually Nicola Spallanzani who contributed most of the text contacted me about this, worried that it was not possible to run with AiiDA) so I feel it is important to put it in the docs ASAP, as many more users will have the exact same question (already some of us developers were worried that AiiDA does not support passphrase-protected keys).

Also, I've been using the script for the past 3 years, and Nicola Spallanzani tried it and confirmed that it works. So I'm quite confident that it is working, or in the worst case can serve as a starting point for users to write their own adaptation.

Anyway, if this is controversial and delaying the merge of the PR, I'll remove it and make an independent PR.

csadorf replied:
Let's leave it in to avoid creating work.

giovannipizzi committed:
Apply suggestions from second round of review

Co-authored-by: Carl Simon Adorf <carl.simon.adorf@gmail.com>

giovannipizzi commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,254 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+  3. Prevent your operating system from indexing the file repository.
+
+     .. dropdown:: Disabling ``locate`` on Linux
+
+        Many Linux distributions include the ``locate`` command to quickly find files and folders, and run a daily cron job ``updatedb.mlocate`` to create the corresponding index.
+        A large file repository can take a long time to index, up to the point where the hard drive is constantly indexing.
+
+        In order to exclude the repository folder from indexing, add its path to the ``PRUNEPATH`` variable in the ``/etc/updatedb.conf`` configuration file (use ``sudo``).
```
This works for me - make sure to have `sphinx-panels~=0.4.0`. I have now rebased so the dependency is in.

giovannipizzi commented:
@csadorf comments addressed (or commented/replied above) - this is now ready for re-review

chrisjsewell commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,327 @@ the configuration directory ``~/project_a/.aiida`` will be used.
+
+Updating from 0.x.* to 1.*
+--------------------------
+- `Additional instructions on how to migrate from 0.12.x versions <https://aiida.readthedocs.io/projects/aiida-core/en/v1.2.1/install/updating_installation.html#updating-from-0-12-to-1>`_.
+- `Additional instructions on how to migrate from versions 0.4 -- 0.11 <https://aiida.readthedocs.io/projects/aiida-core/en/v1.2.1/install/updating_installation.html#older-versions>`_.
+- For a list of breaking changes between the 0.x and the 1.x series of AiiDA, check `this page <https://aiida.readthedocs.io/projects/aiida-core/en/v1.2.1/install/updating_installation.html#breaking-changes-from-0-12-to-1>`_.
```
The other approach to this is to keep move these sections to e.g.  `docs/source/howto/_version_migrations.rst` and link to that.
Then at the top of that document put `:orphan:`; this tells sphinx that the document is not required to show up in any ToC. This is what I have currently done in #4060 (see https://7-77234579-gh.circle-artifacts.com/0/html/intro/installation.html#updating-aiida-core), but I will remove that section and link here

chrisjsewell commented on docs/source/howto/installation.rst:
```diff
@@ -223,24 +223,327 @@ the configuration directory ``~/project_a/.aiida`` will be used.
 Tuning performance
```
```suggestion
.. _how-to:installation:tuning:

Tuning performance
```

chrisjsewell replied:
oops sorry, didn't notice this was above

csadorf reviewed (approved):
Looks good!
